etna/search/forms.py

```diff
@@ -170,56 +201,43 @@ def clean(self):
             # Either one or both date fields are empty. No further validation necessary.
             ...
 
-        cleaned_data["filter_aggregations"] = (
-            [cleaned_data.get("group")]
-            + cleaned_data.get("levels")
-            + cleaned_data.get("topics")
-            + cleaned_data.get("collections")
-            + cleaned_data.get("closure_statuses")
-            + cleaned_data.get("catalogue_sources")
-        )
-
         return cleaned_data
 
-    def update_from_response(self, *, response):
-        """Populate dynamic fields choices using aggregation data from API."""
-        aggregations = response["aggregations"]
-
-        self.fields["levels"].update_from_aggregations(
-            aggregations.get("level", {}).get("buckets")
-        )
-        self.fields["topics"].update_from_aggregations(
-            aggregations.get("topic", {}).get("buckets")
-        )
-        self.fields["collections"].update_from_aggregations(
-            aggregations.get("collection", {}).get("buckets")
-        )
-        self.fields["closure_statuses"].update_from_aggregations(
-            aggregations.get("closure", {}).get("buckets")
-        )
-        self.fields["catalogue_sources"].update_from_aggregations(
-            aggregations.get("catalogueSource", {}).get("buckets")
-        )
-
-    def selected_filters(self):
-        """List of selected filters, keyed by the corresponding field name.
+    @cached_property
+    def selected_filters(self) -> Dict[str, List[Tuple[str, str]]]:
+        """List of selected values, keyed by the corresponding field name.
 
         Used by template to output a list of selected filters.
 
         Method must be called on a bound form (post validation)
 
-        Note: selected filters differ from the filter_aggregations passed to
-        the client as 'group' isn't considered to be a filter.
-
+        TODO: When we inevitably shift to class-based views, this logic
+        should be moved to the view.
         """
-        return {
-            "levels": self.cleaned_data.get("levels"),
-            "topics": self.cleaned_data.get("topics"),
-            "collections": self.cleaned_data.get("collections"),
-            "closure_statuses": self.cleaned_data.get("closure_statuses"),
-            "catalogue_sources": self.cleaned_data.get("catalogue_sources"),
+        return_value = {
+            field_name: self.cleaned_data.get(field_name)
+            for field_name in self.dynamic_choice_fields
         }
 
+        # Remove items with no values
+        return_value = {k: v for k, v in return_value.items() if v}
+
+        # Replace raw values with (key, label) tuples
+        for key, values in return_value.items():
+            for value, label in self.fields[key].choices:
+                if value in values:
+                    label_without_count = re.sub(r" \([0-9\,]+\)", "", label, 0)
```

Comment:
Good spot!

Because all of the filters will have counts added at the end, we can modify the regex slightly to only strip the first instance of the match from the very end of the string - leaving anything else in-tact.